File: scala_stuff/src/cashflow/java/Cashflow.java
package cashflow.java;

import java.time.LocalDate;
import java.time.LocalDateTime;

public class Cashflow {
	private final double amount;
	private final String currency;
	private final LocalDateTime due;

	public Cashflow(double amount, String currency, LocalDateTime due) {
		this.amount = amount;
		this.currency = currency;
		this.due = due;
	}

	public double getAmount() {
		return amount;
	}

	public String getCurrency() {
		return currency;
	}

	public LocalDateTime getDue() {
		return due;
	}
}

File: scala_stuff/src/cashflow/scala/Cashflow.scala
package cashflow.scala

import java.time.{LocalDate, LocalDateTime}
import java.time.temporal.TemporalAmount

class Cashflow(val amount: Double, val currency: String, val due: LocalDateTime){
  def this(amount: Double, due: LocalDateTime) = this(amount, "USD", due)

  def this(amount: Double) = this(amount, "GBP", LocalDateTime.now())

  val settle: LocalDate = due.toLocalDate.plusDays(2)
  private lazy val processedAt = LocalDateTime.now

  def rollForward(): Cashflow = {
    val retval = new Cashflow(amount, currency, due.plusDays(1))
    retval.processedAt
    retval
  }
}

object Cashflow{
  def main(args: Array[String]): Unit = {
    val c1 = new Cashflow(100.23)
    println(c1.currency)
    val c2 = new Cashflow(300.00, LocalDateTime.now)
    Thread.sleep(1000)
    println(c1.settle)
    println(c2.rollForward())
    println(c1.processedAt)
  }
}
